File: app/src/test/java/coolpharaoh/tee/speicher/tea/timer/views/show_tea/countdowntimer/NotifierTest.java
package coolpharaoh.tee.speicher.tea.timer.views.show_tea.countdowntimer;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.when;

import android.app.Application;
import android.app.Notification;
import android.os.Build;

import androidx.test.core.app.ApplicationProvider;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import coolpharaoh.tee.speicher.tea.timer.R;
import coolpharaoh.tee.speicher.tea.timer.core.system.CurrentSdk;
import coolpharaoh.tee.speicher.tea.timer.core.system.SystemUtility;

//could be removed when Robolectric supports Java 8 for API 29
@Config(sdk = Build.VERSION_CODES.O_MR1)
@RunWith(RobolectricTestRunner.class)
public class NotifierTest {
    public static final String CHANNEL_ID_NOTIFY = "3422";
    public static final String TEA_NAME = "Earl Grey";

    @Rule
    public MockitoRule rule = MockitoJUnit.rule();
    @Mock
    TimerViewModel timerViewModel;
    @Mock
    SystemUtility systemUtility;

    @Before
    public void setUp() {
        CurrentSdk.setFixedSystem(systemUtility);
        when(timerViewModel.getName(1L)).thenReturn(TEA_NAME);
    }

    @Test
    public void getNotificationAfterAndroidO() {
        when(systemUtility.getSdkVersion()).thenReturn(Build.VERSION_CODES.O_MR1);

        final Application application = ApplicationProvider.getApplicationContext();
        final Notifier notifier = new Notifier(application, 1L, timerViewModel);
        final Notification notification = notifier.getNotification();

        assertThat(notification.getChannelId()).isEqualTo(CHANNEL_ID_NOTIFY);
        assertThat(notification.tickerText).isEqualTo(application.getString(R.string.show_tea_notification_ticker));
        assertThat(notification.extras.get("android.title")).isEqualTo(application.getString(R.string.show_tea_notification_title));
        assertThat(notification.extras.get("android.text")).isEqualTo(TEA_NAME);
    }

    @Test
    public void getNotificationBeforeAndroidO() {
        when(systemUtility.getSdkVersion()).thenReturn(Build.VERSION_CODES.N_MR1);

        final Application application = ApplicationProvider.getApplicationContext();
        final Notifier notifier = new Notifier(application, 1L, timerViewModel);
        final Notification notification = notifier.getNotification();

        assertThat(notification.getChannelId()).isNull();
        assertThat(notification.tickerText).isEqualTo(application.getString(R.string.show_tea_notification_ticker));
        assertThat(notification.extras.get("android.title")).isEqualTo(application.getString(R.string.show_tea_notification_title));
        assertThat(notification.extras.get("android.text")).isEqualTo(TEA_NAME);
    }
}

File: app/src/test/java/coolpharaoh/tee/speicher/tea/timer/views/show_tea/countdowntimer/NotifierTest.kt
package coolpharaoh.tee.speicher.tea.timer.views.show_tea.countdowntimer

import android.app.Application
import androidx.test.core.app.ApplicationProvider
import coolpharaoh.tee.speicher.tea.timer.R
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit4.MockKRule
import org.assertj.core.api.Assertions.*
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class NotifierTest {
    @get:Rule
    val mockkRule = MockKRule(this)
    @MockK
    lateinit var timerViewModel: TimerViewModel

    @Before
    fun setUp() {
        every { timerViewModel.getName(1L) } returns TEA_NAME
    }

    @Test
    fun getNotificationAfterAndroidO() {
        val application = ApplicationProvider.getApplicationContext<Application>()
        val notifier = Notifier(application, 1L, timerViewModel)
        val notification = notifier.notification

        assertThat(notification.channelId).isEqualTo(CHANNEL_ID_NOTIFY)
        assertThat(notification.tickerText).isEqualTo(application.getString(R.string.show_tea_notification_ticker))
        assertThat(notification.extras["android.title"]).isEqualTo(application.getString(R.string.show_tea_notification_title))
        assertThat(notification.extras["android.text"]).isEqualTo(TEA_NAME)
    }

    companion object {
        const val CHANNEL_ID_NOTIFY = "3422"
        const val TEA_NAME = "Earl Grey"
    }
}